server/src/main/java/de/uftos/repositories/solver/timefold/constraints/ConstraintDefinitionFactory.java

```diff
@@ -0,0 +1,1073 @@
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.LESSON) {
+            resultSet.add(((LessonTimefoldInstance) resource).getRoom());
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "subject" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.LESSON) {
+            resultSet.add(((LessonTimefoldInstance) resource).getSubject());
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "studentGroup" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.LESSON) {
+            resultSet.add(((LessonTimefoldInstance) resource).getStudentGroup());
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "grade" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.STUDENT_GROUP) {
+            resultSet.add(((StudentGroupTimefoldInstance) resource).getGrade());
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "day" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.TIMESLOT) {
+            resultSet.add(new Number(((TimeslotTimefoldInstance) resource).getDayOfWeek()));
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "slot" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.TIMESLOT) {
+            resultSet.add(new Number(((TimeslotTimefoldInstance) resource).getSlotId()));
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "students" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          switch (resource.getResourceType()) {
+            case TIMETABLE ->
+                resultSet.addAll(((TimetableSolutionTimefoldInstance) resource).getStudents());
```

Comment:
`resource` is a single item, right? Why do you need `addAll` then? AM I overlooking something?

Reply:
Without cast it would be `addAll(resource.getStudents())` and `getStudents` will return a List.

Reply:
Oh sorry, I read it as `addAll(resource).getStudents()`